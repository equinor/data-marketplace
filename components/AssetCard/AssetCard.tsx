import { Typography } from "@equinor/eds-core-react"
import { tokens } from "@equinor/eds-tokens"
import Link from "next/link"
import type { VoidFunctionComponent } from "react"
import styled from "styled-components"

const CardContainer = styled.div`
  box-shadow: ${tokens.elevation.raised};
  border-radius: ${tokens.shape.corners.borderRadius};
  padding: 2rem;

  &:not(:last-child) {
    margin-bottom: 1.5rem;
  }

  a > *:not(:last-child) {
    margin-bottom: 1rem;
  }
`

const MetaContainer = styled.div`
  display: flex;

  > *:not(:last-child) {
    margin-right: 1rem;
  }
`

const TruncatedDescription = styled(Typography)<{ lines?: number }>`
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: ${({ lines }) => (lines && lines > 0 ? lines : 4)};
  overflow: hidden;
`

type Props = {
  description: string
  id: string
  meta?: { label: string, value: string|number }[]
  title: string
}

export const AssetCard: VoidFunctionComponent<Props> = ({
  description,
  id,
  meta,
  title,
}) => (
  <CardContainer>
    <Link href={{ pathname: "/assets/[id]", query: { assetId: id } }}>
      {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
      <a>
        <div>
          <Typography variant="h4" as="p">{title}</Typography>
        </div>

        {meta && (
          <MetaContainer>
            {meta.map((m, i) => (
              <Typography variant="caption" key={`${id}-meta-${i + 1}`}>
                {m.label}
                {" "}
                {m.value}
              </Typography>
            ))}
          </MetaContainer>
        )}

        <div>

          {/*
            * html strings are sanitized in backend at this point,
            * so we should(TM) be safe to do this
            */}
          <TruncatedDescription variant="body_long" dangerouslySetInnerHTML={{ __html: description }} />
        </div>
      </a>
    </Link>
  </CardContainer>
)
