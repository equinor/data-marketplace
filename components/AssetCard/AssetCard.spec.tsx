import { render, screen } from "@testing-library/react"

import { AssetCard } from "./AssetCard"

describe("AssetCard", () => {
  it("renders no meta container", () => {
    const { container } = render(<AssetCard id="abc-123" description="some description" title="some title" />)
    expect(screen.queryByTestId("meta-container")).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it("renders meta container", () => {
    const meta = [{ label: "Some label", value: "Some value" }]
    const { container } = render(
      <AssetCard
        id="abc-123"
        meta={meta}
        description="some description"
        title="some title"
      />,
    )
    const metaContainer = screen.queryByTestId("meta-container")
    expect(metaContainer).toBeInTheDocument()
    expect(metaContainer?.children.length).toBe(meta.length)
    expect(container).toMatchSnapshot()
  })
})
