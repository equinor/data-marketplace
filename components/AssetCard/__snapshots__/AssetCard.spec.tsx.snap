// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetCard renders meta container 1`] = `
<div>
  <div
    class="sc-ce74cd1f-0 dhuBPe"
  >
    <a
      href="/assets/abc-123"
    >
      <div>
        <p
          class="Typography__StyledTypography-sc-179guof-0 eza-DtJ"
        >
          some title
        </p>
      </div>
      <div
        class="sc-ce74cd1f-1 YZBsu"
        data-testid="meta-container"
      >
        <p
          class="Typography__StyledTypography-sc-179guof-0 efTvKc"
        >
          Some label
           
          Some value
        </p>
      </div>
      <div>
        <p
          class="Typography__StyledTypography-sc-179guof-0 llQfUJ sc-429d35d-0 gflvhq"
        >
          some description
        </p>
      </div>
    </a>
  </div>
</div>
`;

exports[`AssetCard renders no meta container 1`] = `
<div>
  <div
    class="sc-ce74cd1f-0 dhuBPe"
  >
    <a
      href="/assets/abc-123"
    >
      <div>
        <p
          class="Typography__StyledTypography-sc-179guof-0 eza-DtJ"
        >
          some title
        </p>
      </div>
      <div>
        <p
          class="Typography__StyledTypography-sc-179guof-0 llQfUJ sc-429d35d-0 gflvhq"
        >
          some description
        </p>
      </div>
    </a>
  </div>
</div>
`;
